### PR TITLE
Add a few methods to stream:

### PR DESCRIPTION
**Pull request by alexeieleusis**
## Constructors
- [x] empty
- [x] error
- [x] never
- [x] of

## Methods
- [x] combineLatest
- [x] debounce
- [x] debounceTime
- [x] merge
- [x] scan
- [x] replay

Progress towards #14

### CODE REVIEW

File: lib/src/instances/stream_monad.dart [hunk 1]
```diff
@@ -68,10 +84,109 @@ class StreamMonad<T> extends Monad<T> implements Stream<T> {
   StreamMonad<E> asyncMap<E>(FutureOr convert(T event)) =>
       new StreamMonad(_stream.asyncMap(convert));
 
+  /// Combines multiple Observables to create an Observable whose values are
+  /// calculated from the latest values of each of its input Observables.
+  StreamMonad<E> combineLatest<S, E>(
+      StreamMonad<S> other, E combine(T element, S otherElement)) {
```
Comment (wigahluk) [inline lib/src/instances/stream_monad.dart]: Shouldn't be `combine` optional? of you provide it, it is used, if not, a tuple is created

File: lib/src/instances/stream_monad.dart [hunk 2]
```diff
@@ -68,10 +84,109 @@ class StreamMonad<T> extends Monad<T> implements Stream<T> {
...
+        other.listen(processS, onError: onError, onDone: onDone);
+
+    return new StreamMonad(combined.stream);
+  }
+
   @override
   FutureMonad<bool> contains(Object needle) =>
       new FutureMonad(_stream.contains(needle));
 
+  /// Emits a value from the source Observable only after a particular time span
```
Comment (wigahluk) [inline lib/src/instances/stream_monad.dart]: the comment says the time span is determined by another observable, but here the duration is sent as an argument.

File: lib/src/instances/stream_monad.dart [hunk 3]
```diff
@@ -68,10 +84,109 @@ class StreamMonad<T> extends Monad<T> implements Stream<T> {
...
+  /// determined by another Observable has passed without another source
+  /// emission.
+  StreamMonad<T> debounce(Duration duration) {
+    final controller = new StreamController<T>.broadcast();
```
Comment (wigahluk) [inline lib/src/instances/stream_monad.dart]: Why is always broadcasting? shouldn't be broadcasted only if the source is also broadcasting?

File: lib/src/instances/stream_monad.dart [hunk 4]
```diff
@@ -68,10 +84,109 @@ class StreamMonad<T> extends Monad<T> implements Stream<T> {
...
+      emitted = false;
+    });
+
+    return new StreamMonad(controller.stream);
+  }
+
+  /// Emits a value from the source Observable only after a particular time span
+  /// has passed without another source emission.
+  StreamMonad<T> debounceTime(Duration duration) {
+    final controller = new StreamController<T>.broadcast();
```
Comment (wigahluk) [inline lib/src/instances/stream_monad.dart]: Same question about broadcasting.

File: lib/src/instances/stream_monad.dart [hunk 5]
```diff
@@ -169,6 +280,41 @@ class StreamMonad<T> extends Monad<T> implements Stream<T> {
   @override
   StreamMonad<S> map<S>(Function1<T, S> f) => new StreamMonad(_stream.map(f));
 
+  /// subscribes to each given input Observable (as arguments), and simply
```
Comment (wigahluk) [inline lib/src/instances/stream_monad.dart]: The comment says "to each give..." but there is only one argument

File: lib/src/instances/stream_monad.dart (hunk 2 above)
Comment (alexeieleusis) [reply lib/src/instances/stream_monad.dart]: Fixed.

File: lib/src/instances/stream_monad.dart (hunk 5 above)
Comment (alexeieleusis) [reply lib/src/instances/stream_monad.dart]: Fixed.

File: lib/src/instances/stream_monad.dart (hunk 4 above)
Comment (alexeieleusis) [reply lib/src/instances/stream_monad.dart]: Fixed.

File: lib/src/instances/stream_monad.dart (hunk 3 above)
Comment (alexeieleusis) [reply lib/src/instances/stream_monad.dart]: Fixed, I just dislike single subscription streams.

File: lib/src/instances/stream_monad.dart (hunk 1 above)
Comment (alexeieleusis) [reply lib/src/instances/stream_monad.dart]: Missed this!
Types don't match, we would need to remove the `E` type parameter and instead replace it for `Tuple2<T, S>` I actually like that alternative, consumers can then compose with the combine function if they need to, thoughts?

**Pull request merged**